Applications/Spire/Include/Spire/Ui/FocusObserver.hpp

```diff
@@ -0,0 +1,87 @@
+#ifndef SPIRE_FOCUS_OBSERVER_HPP
+#define SPIRE_FOCUS_OBSERVER_HPP
+#include "Spire/Ui/Ui.hpp"
+
+namespace Spire {
+
+  /** Represents an observer to keep track of the focus states of a widget. */
+  class FocusObserver {
+    public:
+      
+      /**
+       * Specifies the focus state.
+       */
+      enum class State : std::uint8_t {
+
+        /** The state indicates a widget doesn't have focus. */
+        NONE = 0,
+
+        /** The state indicates a widget gets focus. */
+        FOCUS = 0x01,
+
+        /**
+         * The state indicates a widget or one of this widget's children
+         * gets focus.
+         */
+        FOCUS_IN = 0x02,
```

Comment:
These enums model mutually exclusive states, but they have dependencies with one another meaning the definition has the risk of introducing inconsistencies. For example it is never possible to have ```FOCUS``` without ```FOCUS_IN```, hence ```FOCUS``` must include ```FOCUS_IN```, a similar property holds for ```FOCUS_VISIBLE``` something along the lines of:

   FOCUS_IN = 0x01,
   FOCUS = FOCUS_IN | 0x02,
   FOCUS_VISIBLE = FOCUS | 0x04

This way it's never possible to be in the ```FOCUS``` state without being in the ```FOCUS_IN``` state as well.

Clarify whether ```FOCUS_VISIBLE``` requires ```FOCUS``` or just ```FOCUS_IN```.

Reply:
I think `FOCUS_VISIBLE` should be independent of `FOCUS` and `FOCUS_IN`. `FOCUS_VISIBLE` is not only a focus state but also indicates the widget gets focus from a keyboard event, so `FOCUS_VISIBLE` could happen upon `FOCUS` or `FOCUS_IN`. When the widget itself gets focus by keyboard, the focus state will be `FOUCS |  FOCUS_VISIBLE`; When a child widget gets focus by keyboard, the state will be `FOCUS_IN | FOCUS_VISIBLE`.

Reply:
You will need to be very careful about the dependencies between these states. Think through it carefully based on the specs and any details you may need from Jon and then revise the API accordingly.

Reply:
According to the definition of these states, focus_in indicates the widget or any of its descendants has focus, whereas focus indicates only the widget gets focus. The overlap of focus_in and focus is the widget itself gets focus. I think Focus_visible is not mutually exclusive to focus/focus_in because it represents the focus getting from the keyboard but it can’t tell whether this widget gets focus or one of its children gets focus.

I think the enums should be

NONE = 0,
FOCUS = 0x01,
FOCUS_IN = FOCUS | 0x02,
FOCUS_VISIBLE = 0x04

Reply:
Okay, this still has issues, it says that FOCUS and FOCUS_VISIBLE are independent states. It also says that having FOCUS_IN requires having FOCUS.

Go over this very carefully again and then update your proposal.

Reply:
Updated the enums:

NONE = 0,
FOCUS_IN = 0x01,
FOCUS = FOCUS_IN | 0x02,
FOCUS_VISIBLE = FOCUS | 0x04

Reply:
This says FOCUS_IN is an independent state.
FOCUS is a new state that depends on being in the FOCUS_IN state.
FOCUS_VISIBLE is a new state that requires being in the FOCUS state (which implies being in the FOCUS_IN state).

If this is correct then update the API accordingly.